### DISCUSSION TITLE
Can Openthread Leader become a commissioner?

### DISCUSSION THREAD

**Discussion by ksathara** [openthread/openthread#10412] [Q&A] [answered]
Hi, 
I am trying a commissioner start and to join an ED. I have only one leader in the network. I can make him a commissioner and join an ED by using the cli, but when I try to do it via API, the commissioner gets stuck in petitioning. 
Any ideas?
Cheers,
Kaushalya

**Comment by jwhui** [comment]
Yes, a Leader can serve as a Commissioner. Your successful results using the CLI demonstrates this.

Note that the CLI is simply built on top of the OT API and serves as an example of using OT API. I would suggest reviewing the CLI implementation to understand how best to call the OT APIs directly.

https://github.com/openthread/openthread/blob/main/src/cli/cli_commissioner.cpp

**Reply by ksathara** [reply]
Thanks Jonathan. My implementation was based on the cli. 

Now if I implement commissioner start like this, it works. I am not creating a new dataset. but using the active dataset. The cli_cmsn () is a custom cli command that I have created to test.

```
otError cli_cmsn (void *aContext, uint8_t aArgsLength, char *aArgs[]) {
	otError error;
	
	otInstance *instance = openthread_get_default_instance();
	if (instance == NULL) {
		otCliOutputFormat ("get instantace  error");
		return OT_ERROR_FAILED;
	}

	otCommissionerStop (instance);

	error = otCommissionerStart (instance, aStateCallback, aJoinerCallback, NULL);
	if (error != OT_ERROR_NONE) {
		otCliOutputFormat ("commissioner start error %d\n\r", error);
		return (error);
	}

	return (OT_ERROR_NONE);
}

```
But I want to init the dataset and use the new dataset before starting the commissioner. So I did it like this.

```
otError cli_cmsn (void *aContext, uint8_t aArgsLength, char *aArgs[]) {
	otOperationalDataset aDataset;
	otError error;
	
	otInstance *instance = openthread_get_default_instance();

	error = otIp6SetEnabled(instance, false);	// ifconfig down

	error = otThreadSetEnabled (instance, false);	// thread stop

	error = otDatasetCreateNewNetwork(instance, &aDataset);	// create new dataset

	error = otDatasetSetActive(instance, &aDataset);	// active new dataset

	error = otIp6SetEnabled(instance, true);	// ifconfig up

	error = otThreadSetEnabled (instance, true); // thread start

	// spawn a thread to handle wait to connect to network
	k_tid_t my_tid = k_thread_create(&cmsn_thread_data, cmsn_stack_area,
                                 K_THREAD_STACK_SIZEOF(cmsn_stack_area),
                                 cmsn_handler,
                                 NULL, NULL, NULL,
                                 CMSN_THREAD_PRIORITY, 0, K_NO_WAIT);

         return (error);
	}

	return (OT_ERROR_NONE);
}

// thread handler
void cmsn_handler (void *, void *, void *) {
	otError error;
	
	while (ot_connected == true) {     //ot_connected is made true if thread role is either child, router or leader
		k_msleep (500);
	}
	while (ot_connected == false) {   //ot_connected is made false if thread role is either detached or disabled
		k_msleep (500);
	}
	
	otInstance *instance = openthread_get_default_instance();
	otCliOutputFormat ("\n\rdevice role %d", otThreadGetDeviceRole(instance));

	otCommissionerStop (instance);     // stop commissioner just in case 

	error = otCommissionerStart (instance, aStateCallback, aJoinerCallback, NULL);
	}
}

```
I can see the commissioner start happens when the thread role is leader. But it fails to get the petition.

Any clue why it behaves like this?

**Reply by jwhui** [reply]
OpenThread API is not thread safe. OpenThread APIs are designed to be called from a single execution context.

**Reply by ksathara** [reply]
Oh ok. I know this is not an OpenThread question, but what would you advise to do in this case? I have to do a busy wait till thread role changes to router or leader before continuing commissioner start.

**Reply by jwhui** [reply, marked as answer]
Register a callback with [`otSetStateChangedCallback()`](https://openthread.io/reference/group/api-instance#otsetstatechangedcallback) to receive events when states change. You can call OT APIs in the callback.

**Reply by ksathara** [reply]
Great !! it worked. Thanks heaps Jonathan.

**Reply by ksathara** [reply]
Sorry Jonathan, one more question. Is it ok to leave a router/leader permanently in a commissioner role?

**Reply by jwhui** [reply]
Thread only allows at most one active commissioner at any given time. You can leave the commissioner active, but it would not allow any other commissioner to become active.

**Reply by ksathara** [reply]
I can now get end to end commissioning under API control. I want to pass custom data to the commissioner state change cb. I see there is a second parameter as 'void *aContext'. I guess I can use this to pass my data to the cb. How can I implement it?

**Reply by jwhui** [reply]
You can specify the `aContext` as part of `otSetStateChangeCallback()`

**Reply by ksathara** [reply]
I was initially using 
`openthread_set_state_changed_cb(on_thread_state_changed);` 
to  register state change callback, and that works. But it doesn't allow me to pass a user context. It only takes ot_instance as context.

To pass a user context, I am trying to use either otSetStateChangedCallback() or openthread_state_changed_cb_register(). But neither of them seem to register my callback as my callback is not getting invoked when thread state changes.

I dont get any errors as return from  otSetStateChangedCallback() or openthread_state_changed_cb_register().

What could be going wrong?

**Reply by jwhui** [reply]
OpenThread has a build time configuration for the number of callbacks to support.

https://github.com/openthread/openthread/blob/89b54dca2b4b7f5da0f203eaac8a9e048738403f/src/core/config/misc.h#L262

I believe `openthread_set_state_changed_cb()` is a Zephyr-specific implementation and that Zephyr is using one of the supported callback slots.

**Reply by ksathara** [reply]
I have tried one callback register function at a time. So even the limit is one, it should work I guess. 

This is how I used otSetStateChangedCallback ().

```
static void on_thread_state_changed(uint32_t flags, void *context)
{
   struct openthread_context *ot_context = context;
   ...
}

void main() {
  ...
  error = openthread_start(openthread_get_default_context());
  error = otSetStateChangedCallback (openthread_get_default_context(), on_thread_state_changed, &aUserData);
}
```

1. Do you see anything wrong in this?
2. I am puzzled how aUserData is passed to on_thread_state_changed () as the *context seem to carry openthread_context variable.  The openthread_context structure doesn't seem to have any member to carry a user data. How can this be?

Cheers,
Kaushalya

**Reply by jwhui** [reply]
> Do you see anything wrong in this?

Zephyr may be calling `otSetStateChangedCallback` directly. You should make sure `OpenThread` is compiled to support enough state change handlers - see https://github.com/orgs/openthread/discussions/10412#discussioncomment-9836202.

> I am puzzled how aUserData is passed to on_thread_state_changed () as the *context seem to carry openthread_context variable. The openthread_context structure doesn't seem to have any member to carry a user data. How can this be?

Whatever context pointer is passed in to `otSetStateChangedCallback` will be passed back to the callback.
